### PR TITLE
fix(expressions): infer MOD output for mixed types (swev-id: django__django-16082)

### LINKED ISSUE #370: Resolve output_field for mixed-type MOD expressions (DecimalField % IntegerField) — swev-id: django__django-16082
## User Request
Resolve output_field when combining numeric expressions with MOD operator.

Description
When writing a Django expression for a query that does MOD, if the types of the query are different (Decimal and Integer), it doesn't resolve the result to a Decimal type, like it does for other mathematical operators.

Issue Comments:
- Duplicate of #33397.
- I'm pretty sure the changes committed as part of #33397, while closely related, doesn't actually address this case. I think it's best to handle this as a separate ticket.
- Hasn't this been addressed by https://github.com/django/django/pull/15271/files#diff-359ab56295cce36b9597e1d65185f695f271955617b40f43df58ce3fe76fd0c8R492? Removing the marked line will make expressions.tests.ExpressionOperatorTests.test_lefthand_modulo_null fail with a Cannot infer type of '%%' expression involving these types: IntegerField, IntegerField. You must set output_field. Which sounds pretty close to what is described in the issue here. Or what am I missing? :)
- Replying to David Wobrock: Hasn't this been addressed by PR #15271? No, this issue is for mixed numeric types, e.g. DecimalField and IntegerField.

## Research Specification
- Location: django/db/models/expressions.py
- Inference: CombinedExpression._resolve_output_field() uses a registry built from _connector_combinations.
- Current mapping: MOD is covered for same-type numeric (Integer/Integer, Float/Float, Decimal/Decimal) and numeric-with-NULL but **not** for mixed-type numeric pairs.
- Expected policy (consistent with ADD/SUB/MUL/DIV):
  - Integer % Decimal => DecimalField
  - Decimal % Integer => DecimalField
  - Integer % Float => FloatField
  - Float % Integer => FloatField
- Proposed change: In _connector_combinations, add Combinable.MOD to the tuple of operators for “Numeric operations – operands of different type” so existing tuples (Integer, Decimal -> Decimal; Integer, Float -> Float; etc.) apply to MOD.

### Code Change
- File: django/db/models/expressions.py
- Section: _connector_combinations — mixed-type numeric ops
- Change:
  - Before: for connector in (Combinable.ADD, Combinable.SUB, Combinable.MUL, Combinable.DIV)
  - After:  for connector in (Combinable.ADD, Combinable.SUB, Combinable.MUL, Combinable.DIV, Combinable.MOD)

### Tests to Add (tests/expressions/tests.py)
- CombinedExpressionTests:
  1) Mixed Decimal/Integer modulo infers DecimalField both orders.
  2) Integer/Integer modulo infers IntegerField.
  3) Decimal % NULL and NULL % Decimal infer DecimalField (using Value(None)).
  4) Optional: mixed Float/Integer modulo infers FloatField.

### Reproduction (current behavior)
- Construct CombinedExpression with mixed Decimal/Integer modulo; observe FieldError:

```python
from django.db.models import DecimalField, IntegerField, CombinedExpression
from django.db.models.expressions import Combinable

expr = CombinedExpression(DecimalField(), Combinable.MOD, IntegerField())
expr.output_field  # raises FieldError: Cannot infer type of '%%' expression involving these types: DecimalField, IntegerField.
```

This will be validated and documented with observed failure and stack trace during implementation.

---
Reference: swev-id: django__django-16082

### PR DESCRIPTION
**Pull request by casey-brooks**
## Summary
- resolves #370
- update mixed-type connector registry to cover MOD and propagate Decimal output when paired with NULL
- add CombinedExpression modulo tests across Decimal/Integer/Float/NULL

## Problem
CombinedExpression raised `FieldError` when inferring `output_field` for mixed numeric modulo expressions (e.g., DecimalField % IntegerField). Queries that rely on modulo with heterogeneous numeric types could not run without manually specifying `output_field`.

## Reproduction Steps
1. Checkout `django__django-16082`.
2. Run:
   ```bash
   PYTHONPATH=/root/.nix-profile/lib/python3.11/site-packages:$PWD DJANGO_SETTINGS_MODULE=tests.test_sqlite python3 tests/runtests.py expressions --parallel=1
   ```

### Observed Failure (before fix)
```text
ERROR: test_resolve_output_field_modulo (expressions.tests.CombinedExpressionTests.test_resolve_output_field_modulo) (lhs=<class 'django.db.models.fields.DecimalField'>, rhs=<class 'django.db.models.fields.IntegerField'>, combined=<class 'django.db.models.fields.DecimalField'>)
  File "/workspace/django/tests/expressions/tests.py", line 2452, in test_resolve_output_field_modulo
    self.assertIsInstance(expr.output_field, combined)
  File "/workspace/django/django/utils/functional.py", line 57, in __get__
    res = instance.__dict__[self.name] = self.func(instance)
  File "/workspace/django/django/db/models/expressions.py", line 659, in _resolve_output_field
    raise FieldError(
FieldError: Cannot infer type of '%%' expression involving these types: DecimalField, IntegerField. You must set output_field.
```

## Tests
- Added CombinedExpression modulo coverage for Decimal/Integer/Float/NULL cases.
- `PYTHONPATH=/root/.nix-profile/lib/python3.11/site-packages:$PWD DJANGO_SETTINGS_MODULE=tests.test_sqlite python3 tests/runtests.py expressions --parallel=1`

## Environment Adjustments
- `nix profile add nixpkgs#python311`
- `nix profile add nixpkgs#python311Packages.asgiref`
- `nix profile add nixpkgs#python311Packages.sqlparse`
- `nix profile add nixpkgs#python311Packages.pytz`

### CODE REVIEW

**Comment by casey-brooks** [comment]
Test & Lint Summary
- `PYTHONPATH=/root/.nix-profile/lib/python3.11/site-packages:/workspace/django DJANGO_SETTINGS_MODULE=tests.test_sqlite python3 tests/runtests.py expressions --parallel=1` (passed=180, failed=0, skipped=1; expected failures=1)
- `PYTHONPATH=/root/.nix-profile/lib/python3.11/site-packages:/workspace/django flake8 django/db/models/expressions.py tests/expressions/tests.py` (no lint issues)

File: django/db/models/expressions.py [hunk 1]
```diff
@@ -564,6 +565,12 @@ def __hash__(self):
         )
         for field_type in (fields.IntegerField, fields.DecimalField, fields.FloatField)
     },
+    {
+        Combinable.MOD: [
```
Comment (noa-lucent) [inline django/db/models/expressions.py]: [minor] We already register  combinations with  for *all* numeric field types in the loop just above. Adding a dedicated block for the Decimal case duplicates those entries and doesn't change the resolver's behaviour. Let's drop this extra block so the registry stays minimal and avoids suggesting there's a special MOD/NULL rule.

File: tests/expressions/tests.py [hunk 2]
```diff
@@ -2429,6 +2429,44 @@ def test_resolve_output_field_number(self):
...
+            return Expression(value)
+
+        def null():
+            return Value(None)
+
+        numeric_tests = [
+            (DecimalField, IntegerField, DecimalField),
+            (IntegerField, DecimalField, DecimalField),
+            (IntegerField, IntegerField, IntegerField),
+            (FloatField, IntegerField, FloatField),
```
Comment (noa-lucent) [inline tests/expressions/tests.py:2449]: [minor] Thanks for covering the mixed Decimal/Integer cases and guarding the Integer % Integer regression. We also added `(fields.IntegerField, fields.FloatField, fields.FloatField)` to `_connector_combinations` for `MOD`, so we should assert that case here as well (mirroring the Float/Integer check) to catch any future drift. Could you add a subtest for `IntegerField % FloatField` → `FloatField`?

**Review by noa-lucent — CHANGES_REQUESTED** [review]
Thanks for extending the connector table and adding focused coverage. I spotted two small follow-ups before we can merge: 1) the dedicated MOD/None mapping for Decimal is redundant with the existing numeric+NULL registrations, and 2) we should assert the Integer % Float case alongside the other modulo tests. Once those are addressed this should be good to go.

**Comment by casey-brooks** [comment]
Test & Lint Summary (update)
- `PYTHONPATH=/root/.nix-profile/lib/python3.11/site-packages:/workspace/django DJANGO_SETTINGS_MODULE=tests.test_sqlite python3 tests/runtests.py expressions --parallel=1` (passed=180, failed=0, skipped=1; expected failures=1)
- `PYTHONPATH=/root/.nix-profile/lib/python3.11/site-packages:/workspace/django flake8 django/db/models/expressions.py tests/expressions/tests.py` (no lint issues)

**Review by noa-lucent — APPROVED** [review]
Thanks for the prompt fixes—both earlier points are addressed. The connector table now relies solely on the shared numeric/NULL registrations, and the modulo tests cover the Integer%Float path alongside the Decimal scenarios. Looks good to me.
